atmat/atphysics/TouschekPiwinski/MomAperture_Project2Start.m

```diff
@@ -0,0 +1,193 @@
+    eu_ini=EACC;
+end
+
+
+%% bipatition method for multiple points
+orbit=findorbit6(THERING,REFPTS);
+
+% positive branch
+et=et_ini*ones(np,1);
+eu=eu_ini*ones(np,1);
+es=es_ini*ones(np,1);
+de=1;
+iteration=0;
+while de>detole
+    if verbose
+        seconds_initial=datetime('now');
+        fprintf('Positive branch search, iteration %d...',iteration);
+    end
+    L=Multiorigin_ringpass_islost(THERING,REFPTS,et,orbit,nturns,initcoord);
+    es(L==0)=et(L==0);
+    eu(L~=0)=et(L~=0);
+    et=(es+eu)/2;
+    de=max(abs(es-eu));
+    if verbose
+        elapsed_time=seconds(time(between(seconds_initial,datetime('now'))));
+        fprintf('Elapsed time is %1.3f seconds. Energy resolution is %1.3e\n',elapsed_time,de);
+    end
+    iteration=iteration+1;
+end
+etp=et;%+orbit(5,:)';
+etp(etp<1e-6)=1e-6;
+
+% negative branch
+et=-et_ini*ones(np,1);
+eu=-eu_ini*ones(np,1);
+es=-es_ini*ones(np,1);
+de=1;
+iteration=0;
+while de>detole
+    if verbose
+        seconds_initial=datetime('now');
+        fprintf('Negative branch search, iteration %d...',iteration);
+    end
+    L=Multiorigin_ringpass_islost(THERING,REFPTS,et,orbit,nturns,initcoord);
+    es(L==0)=et(L==0);
+    eu(L~=0)=et(L~=0);
+    et=(es+eu)/2;
+    de=max(abs(es-eu));
+    if verbose
+        elapsed_time=seconds(time(between(seconds_initial,datetime('now'))));
+        fprintf('Elapsed time is %1.3f seconds. Energy resolution is %1.3e\n',elapsed_time,de);
+    end
+    iteration=iteration+1;
+end
+etn=et;%+orbit(5,:)';
+etn(etn>-1e-6)=-1e-6;
+end
+
+
+function Loste=Multiorigin_ringpass_islost(THERING,refpts,e,orbit,nturns,initcoord)
+% Returns an boolean array: tells whether the particle nposs launched at
+% position poss(ii) with energy e(ii) is lost or not.
+%
+%   Inputs:
+%       -THERING: cell array Lattice used for the traking.
+%       -refpts: array [npossx1] with the elements number where to start the traking.
+%       -e: array [npossx1] with the relative energy deviation at each refpts.
+%       -orbit: array [npossx6] with the orbit at each refpts.  
+%       -nturns: number of full turns to track.
+%       -initcoord: deviation from the 6D closed orbit, same dor each refpts.
+
+nposs=numel(refpts);
+Loste=ones(nposs,1);
+Rin=zeros(6,nposs);
+Rout=zeros(6,nposs);
+small=0.000001;
+tiny=100*eps;
+
+% first track the remaining portion of the ring
+for ii=1:nposs
+    Line=THERING(refpts(ii):end);
+    Rin(:,ii)=orbit(:,ii)+[initcoord(1) 0 initcoord(2) 0 e(ii) 0.0]';
+    Rout(:,ii)=ringpass(Line,Rin(:,ii));
+    Loste(ii)=sign(sum(isnan(Rout(:,ii))));
+end
+ngood1=nposs-sum(Loste);
+Rin1=Rout(:,Loste==0);
+%Rin1(6,:)=0;
+    
+%Just track particles that have survived to the ring end 
+if ngood1==1
+    [~, LOSS1] =ringpass(THERING,Rin1,nturns);
+    Loste(Loste==0)=LOSS1; 
+elseif ngood1>1
+    % search for non numerically similar (100 x eps) particles
+    DiffR=squeeze(std(repmat(Rin1,[1 1 ngood1])-repmat(reshape(Rin1,[6 1 ngood1]),[1 ngood1 1])));
```

Comment:
The goal is to avoid tracking particles that do not contribute significantly to the boundary search. For example, two particles starting at different reference points separated by a tiny (e.g. 1 mm) drift. We now that although they are indeed two different places along the accelerator, the particle coordinates at the end of the first turn will be too similar.

You are right is saying that it might be lattice dependent, so, I agree with the optional implementation of the particle filtering.